src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs

```diff
@@ -386,6 +386,252 @@ IValue CopyAndUpdateUdt(TupleValue originalValue)
             return ResolvedExpressionKind.InvalidExpr;
         }
 
+        /// <summary>
+        /// Creates a string literal that contains the given values as interpolation arguments.
+        /// The given string is expected to consist of text interspaced with integer constants between curly brackets.
+        /// The content between non-escaped curly brackets is parsed as integer,
+        /// and replaced with the string representation of the interpolation argument at that index.
+        /// Registers the built string with the scope manager.
+        /// </summary>
+        internal static IValue CreateStringLiteral(GenerationContext sharedState, string str, params IValue[] exs)
+        {
+            static (int, int, int) FindNextExpression(string s, int start)
+            {
+                while (true)
+                {
+                    var i = s.IndexOf('{', start);
+                    if (i < 0)
+                    {
+                        return (-1, s.Length, -1);
+                    }
+                    else if ((i == start) || (s[i - 1] != '\\'))
+                    {
+                        var j = s.IndexOf('}', i + 1);
+                        if (j < 0)
+                        {
+                            throw new FormatException("Missing } in interpolated string");
+                        }
+                        var n = int.Parse(s[(i + 1)..j]);
+                        return (i, j + 1, n);
+                    }
+                    start = i + 1;
+                }
+            }
```

Comment:
(I don't know C# and I can raise a false suspicion)
If the string `s` ends with `{` then after the call `var i = s.IndexOf('{', start)` the var `i` will be indexing the last char of `s`.
The subsequent call `var j = s.IndexOf('}', i + 1)` will try to access the `s[i + 1]` (which can be a character outside of the string `s`), and the topic [`IndexOf(Char, Int32)`](https://docs.microsoft.com/en-us/dotnet/api/system.string.indexof?view=net-5.0#System_String_IndexOf_System_Char_System_Int32_) makes me think that the `ArgumentOutOfRangeException` will be thrown. 
If that's the case then I just wanted to make sure that your code handles that case.

Reply:
The idea in general is that rewrite steps (such as the entire QIR emission) can rely on the correctness of the AST; if we didn't do that, then each rewrite step would have to do a huge amount of validation and error handling. I believe it is hence better to assign the responsibility of doing that to the compiler, and each rewrite step in turn has the responsibility to generate valid AST if it does an AST transformation. One tool that would be nice to add (which we haven't had the time to build yet) is a separate tool for AST validation. While such a tool is not (yet) available, I believe it is still fine for rewrite steps to rely on correctness, since the compiler would at least do some proper error handling when a rewrite step throws, such that issues like that can be traced to which step is faulty with a little bit of work.

Reply:
And to make it explicit: The format of an interpolated string is expected to be valid in that each open { is closed, and the content between {} is expected to be a 32-bit integer.

Reply:
ok. 
(I assume that "content between {} is expected to be _unsigned_ 32-bit integer.")

Reply:
Yes.